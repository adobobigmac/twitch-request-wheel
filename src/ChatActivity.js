const fetch = require('node-fetch');

export const ActivityStatus = {
  ACTIVE: 1,
  IDLE: 2,
  DISCONNECTED: 3
};

const MAX_IDLE_TIME_MINUTES = 10;

export default class ChatActivity {
  constructor(channel) {
    this.channel = channel;
    this.lastMessageTimes = {};
    this.getStatus = this.getStatus.bind(this);
  }

  updateLastMessageTime = (user) => {
    this.lastMessageTimes = {
      ...this.lastMessageTimes,
      [user]: Date.now()
    };
  }

  minsSinceLastChatMessage = (user) => {
    return Math.floor((Date.now()-this.lastMessageTimes[user])/60000);
  }

  async getStatus(user) {
    // broadcaster always counts as active
    if(user === this.channel) {
      return ActivityStatus.ACTIVE;
    }

    // sent a chat message in the last X mins?
    if(this.lastMessageTimes[user] && this.minsSinceLastChatMessage(user) < MAX_IDLE_TIME_MINUTES) {
      return ActivityStatus.ACTIVE;
    }

    // check the twitch api to see if they have an active connection to chat.
    // (thank the lord almighty for free open proxy sites. CORB is annoying.)
    // note that results are fairly heavily cached, and the API may break
    // eventually (is undocumented), but this is what twitch themselves
    // uses to display the list of people connected to chat - best we got.
    return fetch(`https://thingproxy.freeboard.io/fetch/https://tmi.twitch.tv/group/user/${this.channel}/chatters`)
      .then(r => r.json())
      .then(res => {
        console.log(res);
        if(!res.chatters) {
          return ActivityStatus.DISCONNECTED; // in case the response doesn't parse goodly
}
        if(res.chatters.vips.includes(user)
           || res.chatters.moderators.includes(user)
           || res.chatters.viewers.includes(user)
           || res.chatters.staff.includes(user)
           || res.chatters.admins.includes(user)
           || res.chatters.global_mods.includes(user)
        ) {
          console.log("IDLEEE")
          return ActivityStatus.IDLE;
        }
        return ActivityStatus.DISCONNECTED;
      })
      .catch((e) => {
        console.log(e);
      })
  }
}
