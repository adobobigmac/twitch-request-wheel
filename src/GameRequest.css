.game-request {
    box-shadow: inset 1px 1px 40px 0 rgba(0, 0, 0, 0.45);
    background-size: cover;
    border-radius: 5px;
    text-align: left;
    text-transform: capitalize;
}

.game-request.chosen {
    background-color: greenyellow;
    color: black;
}

.game-request.pending {
    background-color: steelblue;
}

.game-request-body {
    align-items: center;
    display: flex;
    font-size: 16px;
    justify-content: space-between;
    margin: 0px 15px 7px;
    padding: 4px;
}

.game-request-body .options {
    display: flex;
}

.game-request .lock {
    width: 16px;
    opacity:  0.2;
    padding-right: 8px;
}
.game-request .lock.locked {
    opacity: 1;
}

.game-request .deleteButton {
    background-color: red;
    float: right;
    height: 50%;
    color: white;
    font-size: 11px;
    padding: 3px 5px;
}

.game-request .tooltip {
    display: flex;
    align-items: center;
}

.game-request .tooltipText {
    margin: 0;
}

.game-request .status {
    width: 8px;
    height: 8px;
    border-radius: 5px;
    margin-left: 6px;
    margin-top: 3px;
}

.game-request .status.active {
    background-color: green;
}

.game-request .status.idle {
    background-color: yellow;
}

.game-request .status.disconnected {
    background-color: darkgrey;
}
