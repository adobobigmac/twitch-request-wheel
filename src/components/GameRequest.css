.game-request {
    box-shadow: inset 1px 1px 40px 0 rgba(0, 0, 0, 0.45);
    background-size: cover;
    border-radius: 5px;
    text-align: left;
    text-transform: capitalize;
}

.game-request.chosen {
    background-color: greenyellow;
    color: black;
}

.game-request.pending {
    background-color: steelblue;
}

.game-request-body {
    align-items: center;
    display: flex;
    font-size: 16px;
    justify-content: space-between;
    margin: 0px 15px 7px;
    padding: 2px;
}

.game-request-body .options {
    display: flex;
}

.game-request .lock {
    width: 24px;
    opacity:  0.2;
    padding-right: 8px;
}

.game-request .deleteButton {
    background-color: red;
    float: right;
    height: 50%;
    color: white;
    font-size: 11px;
    padding: 3px 5px;
    opacity: 0.5;
}
.game-request .lock,
.game-request .deleteButton {
    cursor: pointer;
}
.game-request .lock:hover,
.game-request .lock:active,
.game-request .lock.locked,
.game-request .deleteButton:hover,
.game-request .deleteButton:active {
    opacity: 1;
}


#game-request-tooltip .tooltip-inner {
    max-width: 500px;
}

.game-request .tooltip {
    display: flex;
    align-items: center;
}

.game-request .tooltipText {
    margin: 0;
}

.game-request .status {
    width: 8px;
    height: 8px;
    border-radius: 5px;
    margin-left: 6px;
    margin-top: 3px;
}

.game-request .status.active {
    background-color: green;
}

.game-request .status.idle {
    background-color: yellow;
}

.game-request .status.disconnected {
    background-color: darkgrey;
}
