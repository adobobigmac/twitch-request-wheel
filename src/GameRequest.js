import { Component } from 'react';
import ReactTooltip from 'react-tooltip'
import { ActivityStatus } from './ChatActivity';
import './messageStyles.css';
import lock from './lock.svg';

export default class GameRequest extends Component {
  constructor(props){
    super(props);
    this.state = {
      timeDiff: 0,
      activityStatus: null
    };
  }

  getFormattedTimeDiff = (timestamp) => {
    let unit = "minute";
    let timeDiff = Math.floor((Date.now()-timestamp)/60000);
    if(timeDiff === 0) {
      return "just now";
    }
    if(timeDiff >= 60) {
      unit = "hour";
      timeDiff = Math.floor(timeDiff/60);

      if(timeDiff >= 24) {
        unit = "day";
        timeDiff = Math.floor(timeDiff/24);
      }
    }
    return `${timeDiff} ${unit}${timeDiff === 1 ? "" : "s"} ago`;
  }

  updateStatus = () => {
    // return Date.now();
    this.props.getActivity(this.props.metadata.username).then((activityStatus) => {
      this.setState(() => {
        return {
          ...this.state,
          timeDiff: this.getFormattedTimeDiff(this.props.metadata.time),
          activityStatus
        };
      });
    })
  }
  delete = (e) => {
    this.props.onDelete(this.props.msg)
  }

  toggleLock = () => {
    this.props.toggleLock(this.props.msg)
  }

  getTooltipContents = () => {
    let statusClass = "";
    switch(this.state.activityStatus) {
      case ActivityStatus.ACTIVE:
        statusClass = "active";
        break;
      case ActivityStatus.IDLE:
        statusClass = "idle";
        break;
      case ActivityStatus.DISCONNECTED:
        statusClass = "disconnected";
        break;
      default:
        // no data back yet; don't show an activity status indicator at all
        break;
    }

    return (`
      <div class="tooltip">
        <p class="tooltipText">Requested ${this.state.timeDiff} by ${this.props.metadata.username}</p>
        <div class="status ${statusClass}" />
      </div>`);
  }

  render() {
    const lockOpacity = this.props.metadata.locked ? '1' : '0.2';
    return (
      <div>
      <ReactTooltip effect="solid" place="right"/>
      <div
        id="baseDiv"
        data-tip={this.getTooltipContents()}
        data-html={true}
        style={{
          backgroundColor:'steelblue',
          textAlign: 'left',
          borderRadius: '8px',
          textTransform: 'capitalize'
        }}
        onMouseEnter={this.updateStatus}
      >
      <div style={{margin: '0px 15px 7px', padding: '4px', fontSize: '16px', display: 'flex', justifyContent: 'space-between', alignItems: 'center'}}>
        {this.props.msg}
        <div style={{display:'flex'}}>
          <img src={lock} alt="lock" style={{width: '16px', opacity: lockOpacity, paddingRight: '8px'}} className="lock" onClick={this.toggleLock} />
          <button type='button' className="deleteButton" onClick={this.delete}>X</button>
        </div>
      </div>
      </div>
      </div>
    );
  }
}
