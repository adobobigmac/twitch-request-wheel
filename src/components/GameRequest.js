import {ActivityStatus} from '../ChatActivity';
import React, {Component} from 'react';
import {OverlayTrigger, Tooltip} from 'react-bootstrap'
import lock from '../images/lock.svg';

import './GameRequest.css';

class GameRequest extends Component {
    constructor(props) {
        super(props);
        this.state = {
            timeDiff: 0,
            activityStatus: null
        };

        this.deleteRequest = this.deleteRequest.bind(this);
        this.getFormattedTimeDiff = this.getFormattedTimeDiff.bind(this);
        this.getTooltipContents = this.getTooltipContents.bind(this);
        this.toggleLock = this.toggleLock.bind(this);
        this.updateStatus = this.updateStatus.bind(this);
    }

    deleteRequest = (e) => {
        this.props.onDelete(this.props.gameName)
    }

    getFormattedTimeDiff = (timestamp) => {
        let unit = "minute";
        let timeDiff = Math.floor((Date.now() - timestamp) / 60000);
        if (timeDiff === 0) {
            return "just now";
        }
        if (timeDiff >= 60) {
            unit = "hour";
            timeDiff = Math.floor(timeDiff / 60);

            if (timeDiff >= 24) {
                unit = "day";
                timeDiff = Math.floor(timeDiff / 24);
            }
        }
        return `${timeDiff} ${unit}${timeDiff === 1 ? "" : "s"} ago`;
    }

    getTooltipContents = () => {
        let statusClass = "";
        switch (this.state.activityStatus) {
            case ActivityStatus.ACTIVE:
                statusClass = "active";
                break;
            case ActivityStatus.IDLE:
                statusClass = "idle";
                break;
            case ActivityStatus.DISCONNECTED:
                statusClass = "disconnected";
                break;
            default:
                // no data back yet; don't show an activity status indicator at all
                break;
        }

        // return (`
        //     <div class="tooltip">
        //         <p class="tooltipText">
        //             Requested ${this.state.timeDiff} by ${this.props.metadata.username}
        //         </p>
        //         <div class="status ${statusClass}" />
        //     </div>`
        // );
        return (
            <div className="tooltip">
                <p className="tooltipText">
                    Requested {this.state.timeDiff} by {this.props.metadata.username}
                </p>
                <div className={`status ${statusClass}`} />
            </div>
        );
    }

    toggleLock = () => {
        this.props.toggleLock(this.props.gameName)
    }

    updateStatus = async () => {
        // return Date.now();
        let activityStatus;
        try {
            activityStatus = await this.props.getActivity(this.props.metadata.username);
            return this.setState({
                timeDiff: this.getFormattedTimeDiff(this.props.metadata.time),
                activityStatus
            });
        } catch(err) {
            throw err;
        }
    }

    render() {
        const lockClassName = this.props.metadata.locked ? 'lock locked' : 'lock unlocked';
        const cardStatus = this.props.metadata.chosen ? 'chosen' : 'pending';

        const renderTooltip = (props) => (
            <Tooltip id="game-request-tooltip" {...props}>
                <span className="tooltipText">Requested {this.state.timeDiff} by {this.props.metadata.username}</span>
            </Tooltip>
        );

        return (
            <OverlayTrigger
                placement="left"
                overlay={renderTooltip}
                onEnter={this.updateStatus}>
                <div className="game-request-wrapper fade-in">
                	<div id="baseDiv" className={`game-request ${cardStatus}`}>
                		<div className="game-request-body">
                		    {this.props.gameName}
                			<div className="options">
                				<img src={lock} alt="lock" className={lockClassName} onClick={this.toggleLock} />
                				<button type='button' className="deleteButton" onClick={this.deleteRequest}>X</button>
                			</div>
                		</div>
                	</div>
                </div>
            </OverlayTrigger>
        );
    }
}
export default GameRequest;
