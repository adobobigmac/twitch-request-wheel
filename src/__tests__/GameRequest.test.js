import {ActivityStatus} from '../ChatActivity';
import {createRenderer} from 'react-test-renderer/shallow';
import GameRequest from '../GameRequest';
import React from 'react';
import { render, screen, waitFor } from '@testing-library/react';

// forces react-tooltip to reuse the same uuids
// to avoid breaking snapshots each time :P
jest.mock('crypto', () => ({
    randomBytes: (num: number) => new Array(num).fill(0),
}));

describe('GameRequest', () => {
    const propsQuip = {
        gameName: 'Quiplash',
        metadata: {
            chosen: true,
            locked: true,
            time: Date.now()-300000,
            username: 'Sir Goosewell'
        }
    };
    const propsSTI = {
    	gameName: 'Survive The Internet (Party Pack 4)',
    	metadata: {
    		name: 'Survive The Internet',
    		longName: 'Survive The Internet (Party Pack 4)',
    		partyPack: 'Party Pack 4',
    		'Min players': 3,
    		'Max players': 8,
    		Variants: ['survive', 'survive the internet', 'sti'],
    		username: 'dcpesses',
    		time: Date.now()-120000,
    		locked: false,
    		chosen: false
    	}
    };
    describe('getFormattedTimeDiff', () => {
        test('should return the expected response', async () => {
            let component = new GameRequest({
                metadata: {
                    chosen: true,
                    locked: true,
                    time: Date.now(),
                    username: 'Sir Goosewell'
                }
            });
            expect(component.getFormattedTimeDiff(Date.now())).toBe('just now');
            expect(component.getFormattedTimeDiff(Date.now()-(10 * 60000))).toBe('10 minutes ago');
            expect(component.getFormattedTimeDiff(Date.now()-(60 * 60000))).toBe('1 hour ago');
            expect(component.getFormattedTimeDiff(Date.now()-(24 * 60 * 60000))).toBe('1 day ago');
        });
    });
    describe('render', () => {
        test('should render component using quiplash data', async () => {
            let {container} = render(<GameRequest {...propsQuip} />);
            try {
                await waitFor(() => screen.getByText(/Quiplash/i));
                const element = screen.getByText(/Quiplash/i);
                expect(element).toBeInTheDocument();
            } catch (e) {
                throw e;
            }
            expect(container).toMatchSnapshot();
        });
        test('should render component using survive the internet data', () => {
            const shallowRenderer = createRenderer();
            shallowRenderer.render(<GameRequest {...propsSTI} />);
            let instance = shallowRenderer.getMountedInstance();
            instance.setState({
                activityStatus: ActivityStatus.ACTIVE,
                timeDiff: '2 minutes ago'
            });
            expect(shallowRenderer.getRenderOutput()).toMatchSnapshot();
            instance.setState({
                activityStatus: ActivityStatus.IDLE,
                timeDiff: '3 minutes ago'
            });
            expect(shallowRenderer.getRenderOutput()).toMatchSnapshot();
            instance.setState({
                activityStatus: ActivityStatus.DISCONNECTED,
                timeDiff: '4 minutes ago'
            });
            expect(shallowRenderer.getRenderOutput()).toMatchSnapshot();
            shallowRenderer.unmount();
        });
        test('should render component using survive the internet data and updated status info', async () => {
            let props = Object.assign({}, propsSTI, {
                getActivity: jest.fn()
                    .mockResolvedValueOnce(3)
                    .mockResolvedValueOnce(2)
                    .mockResolvedValueOnce(1),
                onDelete: jest.fn(),
                toggleLock: jest.fn()
            })
            const shallowRenderer = createRenderer();
            shallowRenderer.render(<GameRequest {...props} />);
            let instance = shallowRenderer.getMountedInstance();
            instance.setState({
                activityStatus: ActivityStatus.IDLE,
                timeDiff: '2 minutes ago'
            });
            let component = shallowRenderer.getRenderOutput();
            await component.props.children[1].props.onMouseEnter();
            expect(props.getActivity).toHaveBeenCalledTimes(1);

            // Get the children of the element div.options
            let optionsChildren = component.props.children[1].props.children.props.children[1].props.children;
            optionsChildren[0].props.onClick();
            optionsChildren[0].props.onClick();
            optionsChildren[1].props.onClick();
            expect(props.toggleLock).toHaveBeenCalledTimes(2);
            expect(props.onDelete).toHaveBeenCalledTimes(1);
            shallowRenderer.unmount();
        });
    });
});
