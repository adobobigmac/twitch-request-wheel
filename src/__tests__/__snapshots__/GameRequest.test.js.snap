// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`GameRequest render should render component using quiplash data 1`] = `
<div>
  <div
    class="game-request-wrapper fade-in"
  >
    <div
      class="__react_component_tooltip t00000000-0000-4000-8000-000000000000 place-left type-dark"
      data-id="tooltip"
      id="t00000000-0000-4000-8000-000000000000"
    >
      <style
        aria-hidden="true"
      >
        
  	.t00000000-0000-4000-8000-000000000000 {
	    color: #fff;
	    background: #222;
	    border: 1px solid transparent;
  	}

  	.t00000000-0000-4000-8000-000000000000.place-top {
        margin-top: -10px;
    }
    .t00000000-0000-4000-8000-000000000000.place-top::before {
        border-top: 8px solid transparent;
    }
    .t00000000-0000-4000-8000-000000000000.place-top::after {
        border-left: 8px solid transparent;
        border-right: 8px solid transparent;
        bottom: -6px;
        left: 50%;
        margin-left: -8px;
        border-top-color: #222;
        border-top-style: solid;
        border-top-width: 6px;
    }

    .t00000000-0000-4000-8000-000000000000.place-bottom {
        margin-top: 10px;
    }
    .t00000000-0000-4000-8000-000000000000.place-bottom::before {
        border-bottom: 8px solid transparent;
    }
    .t00000000-0000-4000-8000-000000000000.place-bottom::after {
        border-left: 8px solid transparent;
        border-right: 8px solid transparent;
        top: -6px;
        left: 50%;
        margin-left: -8px;
        border-bottom-color: #222;
        border-bottom-style: solid;
        border-bottom-width: 6px;
    }

    .t00000000-0000-4000-8000-000000000000.place-left {
        margin-left: -10px;
    }
    .t00000000-0000-4000-8000-000000000000.place-left::before {
        border-left: 8px solid transparent;
    }
    .t00000000-0000-4000-8000-000000000000.place-left::after {
        border-top: 5px solid transparent;
        border-bottom: 5px solid transparent;
        right: -6px;
        top: 50%;
        margin-top: -4px;
        border-left-color: #222;
        border-left-style: solid;
        border-left-width: 6px;
    }

    .t00000000-0000-4000-8000-000000000000.place-right {
        margin-left: 10px;
    }
    .t00000000-0000-4000-8000-000000000000.place-right::before {
        border-right: 8px solid transparent;
    }
    .t00000000-0000-4000-8000-000000000000.place-right::after {
        border-top: 5px solid transparent;
        border-bottom: 5px solid transparent;
        left: -6px;
        top: 50%;
        margin-top: -4px;
        border-right-color: #222;
        border-right-style: solid;
        border-right-width: 6px;
    }
  
      </style>
    </div>
    <div
      class="game-request chosen"
      currentitem="false"
      data-html="true"
      data-tip="
            <div class=\\"tooltip\\">
                <p class=\\"tooltipText\\">
                    Requested 0 by Sir Goosewell
                </p>
                <div class=\\"status \\" />
            </div>"
      id="baseDiv"
    >
      <div
        class="game-request-body"
      >
        Quiplash
        <div
          class="options"
        >
          <img
            alt="lock"
            class="lock locked"
            src="file-stub"
          />
          <button
            class="deleteButton"
            type="button"
          >
            X
          </button>
        </div>
      </div>
    </div>
  </div>
</div>
`;

exports[`GameRequest render should render component using survive the internet data 1`] = `
<div
  className="game-request-wrapper fade-in"
>
  <ReactTooltip
    clickable={false}
    effect="solid"
    insecure={true}
    place="left"
    resizeHide={true}
    wrapper="div"
  />
  <div
    className="game-request pending"
    data-html={true}
    data-tip="
            <div class=\\"tooltip\\">
                <p class=\\"tooltipText\\">
                    Requested 2 minutes ago by dcpesses
                </p>
                <div class=\\"status active\\" />
            </div>"
    id="baseDiv"
    onMouseEnter={[Function]}
  >
    <div
      className="game-request-body"
    >
      Survive The Internet (Party Pack 4)
      <div
        className="options"
      >
        <img
          alt="lock"
          className="lock unlocked"
          onClick={[Function]}
          src="file-stub"
        />
        <button
          className="deleteButton"
          onClick={[Function]}
          type="button"
        >
          X
        </button>
      </div>
    </div>
  </div>
</div>
`;

exports[`GameRequest render should render component using survive the internet data 2`] = `
<div
  className="game-request-wrapper fade-in"
>
  <ReactTooltip
    clickable={false}
    effect="solid"
    insecure={true}
    place="left"
    resizeHide={true}
    wrapper="div"
  />
  <div
    className="game-request pending"
    data-html={true}
    data-tip="
            <div class=\\"tooltip\\">
                <p class=\\"tooltipText\\">
                    Requested 3 minutes ago by dcpesses
                </p>
                <div class=\\"status idle\\" />
            </div>"
    id="baseDiv"
    onMouseEnter={[Function]}
  >
    <div
      className="game-request-body"
    >
      Survive The Internet (Party Pack 4)
      <div
        className="options"
      >
        <img
          alt="lock"
          className="lock unlocked"
          onClick={[Function]}
          src="file-stub"
        />
        <button
          className="deleteButton"
          onClick={[Function]}
          type="button"
        >
          X
        </button>
      </div>
    </div>
  </div>
</div>
`;

exports[`GameRequest render should render component using survive the internet data 3`] = `
<div
  className="game-request-wrapper fade-in"
>
  <ReactTooltip
    clickable={false}
    effect="solid"
    insecure={true}
    place="left"
    resizeHide={true}
    wrapper="div"
  />
  <div
    className="game-request pending"
    data-html={true}
    data-tip="
            <div class=\\"tooltip\\">
                <p class=\\"tooltipText\\">
                    Requested 4 minutes ago by dcpesses
                </p>
                <div class=\\"status disconnected\\" />
            </div>"
    id="baseDiv"
    onMouseEnter={[Function]}
  >
    <div
      className="game-request-body"
    >
      Survive The Internet (Party Pack 4)
      <div
        className="options"
      >
        <img
          alt="lock"
          className="lock unlocked"
          onClick={[Function]}
          src="file-stub"
        />
        <button
          className="deleteButton"
          onClick={[Function]}
          type="button"
        >
          X
        </button>
      </div>
    </div>
  </div>
</div>
`;
